### ISSUE TITLE
[no-unnecessary-condition] optional properties should be considered as maybe falsy

### ISSUE THREAD

**Issue by joe223** [typescript-eslint/typescript-eslint#4298]
- [ ] I have tried restarting my IDE and the issue persists.
- [ ] I have updated to the latest version of the packages.
- [ ] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/main/docs/linting/TROUBLESHOOTING.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "warn"
  }
}
```

```TS
interface TrimUserData {
   userInfo?:  {
       id: number
   }
}
const userInfo = _.mapValues(rawUserInfo, (value: TrimUserData) => {
    const info = value.userInfo;
          return {
                id: info?.id // should not throw warning
                // ...
}
```

**Expected Result**
`TrimUserData.userInfo` is optional, so `?.` is necessary when we return 
```
return {
    id: info?.id
    // ...
}
```

eslint should not throw warning or error while using `info?.id`, and should not turn it to `info.id` if running ESLint fix command either.
**Actual Result**

We got an warning which tell us should using `info.id` instead of `info?.id`
```
 id: info?.id
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.31.0` |
| `@typescript-eslint/parser`        | `4.31.0` |
| `TypeScript`                       | `4.4.4` |
| `ESLint`                           | `7.27.0` |
| `node`                             | `14.17.0` |

**Comment by bradzacher** [comment]
Could you please provide more information?
What part is throwing an error? what error are you seeing? Why shouldn't it throw an error?

**Comment by joe223** [comment]
> Could you please provide more information? What part is throwing an error? what error are you seeing? Why shouldn't it throw an error?

I'v updated the description

**Comment by armano2** [comment]
i'm unable to reproduce this with your example - [playground](https://typescript-eslint.io/play/#ts=4.5.2&sourceType=module&code=JYWwDg9gTgLgBAfTgMyhEcDkAbCATAQwGcALTAWACgqBjCAOyPigIHcBVIgUygEl7kEAFxwA3nGB4R9AK4gARjzgAfODPp4uyYPS544AXwDaAXTgBeOKapUA9LbgBaZy9dvXNyjpg9kBGlxwACpQoJw8ACIEMARiVHAJMtx8AhAA-CJxlAk5CZLScopQ8QkGVGXUlHSM8Ek8-IIWiAB0IARgAGoE2DJcRAAULBzJDRAANHD9AG7dvSIhYclRMQCUFgB8WbnVTBKpTTM9XM11KYIA3CUJUFwwMlD0W7k5+XuCac2ScPZwpBAy2H09Ag8BgJDQrDgrAIDx0AHMroZyitPD93Oj0Z5vL5-IEgn0YAAmJ5qEapTKiREvKRwWQKHhXCoVWgMXanUbEywIVrtLpHAZDcJncaTQ5zYIEwlrcybSnZHI7eA6RqWMXHdmpS7y663e6POXPakiZXpT76H5-AFAkFwMEQqEw+jwxFMlGUIA&rules=N4IgAgLgngDgpgZwMYCcCWMIFpEBs0B2EA9AQPZYCuBBcSiCAhilFkmQQCZoRocgAuEHBQoyKEAF8gA&tsConfig=N4IgzgLgTglgxhEAuaBXApgGhOsBZAewBNUAbdASQDsJ0oCAHZNdAXyA)

in your example you provided code like this, that doesn't error on this rule:

```ts
interface TrimUserData {
   userInfo?:  {
       id: number
   }
}

const userInfo = _.mapValues(rawUserInfo, (value: TrimUserData) => {
    const info = value.userInfo;
    return {
        id: info?.id // should not throw warning
    }
})
```

---------------------

if i modify it a little, it reports

```ts
interface Test2 {
   userInfo:  {
       id: number
   }
}

const userInfo2 = _.mapValues(rawUserInfo, (value: Test2) => {
    const info = value.userInfo;
    return {
        id: info?.id // should not throw warning
    }
})
```

**Comment by joe223** [comment]
@armano2 well, I guess this issue is got resolved in later version
